README.md
# unrust / uni-gl

[![Build Status](https://travis-ci.org/unrust/uni-gl.svg?branch=master)](https://travis-ci.org/unrust/uni-gl)
[![Documentation](https://docs.rs/uni-gl/badge.svg)](https://docs.rs/uni-gl)
[![crates.io](https://meritbadge.herokuapp.com/uni-gl)](https://crates.io/crates/uni-gl)

This library is a part of [Unrust](https://github.com/unrust/unrust), a pure rust native/wasm game engine.
This library provides a native/wasm compatibility layer for following components :
* OpenGL API

When used in conjonction with uni-app, on native target, it provides an OpenGL 3.2+ or OpenGLES 2.0+ Core Profile context.
On web target, it provides a WebGL 2.0 context where available, else a WebGL 1.0 context.

## Usage

```toml
[dependencies]
uni-app="0.3.*"
uni-gl="0.3.*"
```

```rust
extern crate uni_app;
extern crate uni_gl;

fn main() {
    // create the game window (native) or canvas (web)
    let app = uni_app::App::new(uni_app::AppConfig {
        size: (800, 600),
        title: "my game".to_owned(),
        vsync: true,
        show_cursor: true,
        headless: false,
        resizable: true,
        fullscreen: false,
        intercept_close_request: false,
    });
    // retrieve the opengl context
    let gl = uni_gl::WebGLRenderingContext::new(app.canvas());
    // start game loop
    app.run(move |_app: &mut uni_app::App| {
        // do some openGL stuff
        gl.clear_color(0.0, 0.0, 1.0, 1.0);
        gl.clear(uni_gl::BufferBit::Color);
    });
}
```

## Build

### As web app (wasm32-unknown-unknown)

Install wasm32 target :
```
rustup target install wasm32-unknown-unknown
```
Install [wasm-pack](https://rustwasm.github.io/wasm-pack/installer/)
and [npm](https://www.npmjs.com/get-npm)

Compile the demo with
```
wasm-pack build examples
```
This creates a wasm package in examples/pkg

Run the demo with
```
cd www
npm install
npm run start
```

Open your browser at http://localhost:8080/

### As desktop app (native-opengl)

```
cargo run --example basic --release
```

## License

Licensed under either of

 * Apache License, Version 2.0, ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)

at your option.

### Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall be dual licensed as above, without any
additional terms or conditions.


CHANGELOG.md
# Changelog

## [0.3.0] - Unreleased
- uses uni-app 0.3.*

## [0.2.1] - 2022-10-11
### Fixed
- missing disable function on webgl

## [0.2.0] - 2022-10-10
### Changed
- upgrade to rust edition 2021
- replace unmaintained cargo-web/stdweb with wasm_bindgen/web-sys

## [0.1.3] - 2020-02-15
### Changed
- upgrade to gl 0.14
- upgrade to stdweb 0.4.20

## [0.1.2] - 2019-12-04
### Fixed
- fixed discrepancy between native and webgl version (don't clear the drawbuffer between frames)

## [0.1.1] - 2019-02-01
### Fixed
- replaced invocation of broken js_raw macro with slower but working js macro

## [0.1.0] - 2018-06-24
- initial release

src/lib.rs
#![recursion_limit = "512"]

#[cfg(not(target_arch = "wasm32"))]
extern crate gl;

#[cfg(target_arch = "wasm32")]
#[path = "webgl.rs"]
pub mod webgl;

#[cfg(not(target_arch = "wasm32"))]
#[path = "webgl_native.rs"]
mod webgl;

#[cfg(not(target_arch = "wasm32"))]
/// whether current OpenGL context is OpenGL ES (Embedded System)
pub const IS_GL_ES: bool = false;

#[cfg(target_arch = "wasm32")]
pub const IS_GL_ES: bool = true;

mod glenum;

pub use glenum::*;
pub use webgl::{GLContext, WebGLContext};

pub mod common {
    use std::ops::{Deref, DerefMut};

    type Reference = super::webgl::Reference;
    type GLContext = super::GLContext;

    #[derive(Debug, Clone)]
    /// The OpenGL rendering context. This is the struct providing most of the OpenGL API.
    pub struct WebGLRenderingContext {
        pub common: GLContext,
    }

    impl Deref for WebGLRenderingContext {
        type Target = GLContext;
        fn deref(&self) -> &GLContext {
            &self.common
        }
    }

    impl DerefMut for WebGLRenderingContext {
        fn deref_mut(&mut self) -> &mut GLContext {
            &mut self.common
        }
    }

    #[derive(Debug)]
    /// an OpenGL buffer created with [`GLContext::create_buffer`].
    ///
    /// Buffers are used to store vertex attributes
    /// (position, normal, uv coordinates) and indexes for indexed rendering,
    pub struct WebGLBuffer(pub Reference);

    impl Deref for WebGLBuffer {
        type Target = Reference;
        fn deref(&self) -> &Self::Target {
            &self.0
        }
    }

    #[derive(Debug)]
    /// an OpenGL shader created with [`GLContext::create_shader`]
    pub struct WebGLShader(pub Reference);
    impl Deref for WebGLShader {
        type Target = Reference;
        fn deref(&self) -> &Self::Target {
            &self.0
        }
    }

    #[derive(Debug, PartialEq)]
    /// an OpenGL shader created with [`GLContext::create_shader`].
    ///
    /// There are two kinds of shaders ([`ShaderKind`]) : vertex and fragment
    pub struct WebGLProgram(pub Reference);
    impl Deref for WebGLProgram {
        type Target = Reference;
        fn deref(&self) -> &Self::Target {
            &self.0
        }
    }

    #[derive(Debug, PartialEq)]
    /// an OpenGL program created with [`GLContext::create_program`].
    ///
    /// It is built with a vertex shader and a fragment shader.
    pub struct WebGLTexture(pub Reference);
    impl Deref for WebGLTexture {
        type Target = Reference;
        fn deref(&self) -> &Self::Target {
            &self.0
        }
    }

    #[derive(Debug)]
    /// an OpenGL vertex array object created with [`GLContext::create_vertex_array`].
    ///
    /// Vertex array objects store all the state needed to supply vertex data.
    pub struct WebGLVertexArray(pub Reference);
    impl Deref for WebGLVertexArray {
        type Target = Reference;
        fn deref(&self) -> &Self::Target {
            &self.0
        }
    }

    #[derive(Debug, PartialEq)]
    /// the reference to a uniform (global GLSL variable) inside a shader, obtained with [`GLContext::get_uniform_location`].
    pub struct WebGLUniformLocation {
        pub reference: Reference,
        pub name: String,
    }
    impl Deref for WebGLUniformLocation {
        type Target = Reference;
        fn deref(&self) -> &Reference {
            &self.reference
        }
    }

    #[derive(Debug)]
    /// an OpenGL Framebuffer created with [`GLContext::create_framebuffer`].
    ///
    /// This is a special type of buffer that can be used as destination for rendering.
    pub struct WebGLFrameBuffer(pub Reference);
    impl Deref for WebGLFrameBuffer {
        type Target = Reference;
        fn deref(&self) -> &Self::Target {
            &self.0
        }
    }

    /// Utility function to print messages to stdout (native) or the js console (web)
    pub fn print(s: &str) {
        GLContext::print(s);
    }
}

pub use self::common::*;


examples/basic.rs
extern crate uni_app;
extern crate uni_gl;

use std::mem::size_of;

#[cfg(target_arch = "wasm32")]
use wasm_bindgen::prelude::*;

#[cfg(target_arch = "wasm32")]
#[wasm_bindgen(start)]
pub fn main_js() -> Result<(), JsValue> {
    main();
    Ok(())
}

// helper to easily convert rust vectors into &[u8] needed by opengl
trait IntoBytes {
    fn into_bytes(self) -> Vec<u8>;
}

impl<T> IntoBytes for Vec<T> {
    fn into_bytes(self) -> Vec<u8> {
        let len = size_of::<T>() * self.len();
        unsafe {
            let slice = self.into_boxed_slice();
            Vec::<u8>::from_raw_parts(Box::into_raw(slice) as _, len, len)
        }
    }
}

fn main() {
    // create the game window (native) or canvas (web)
    let app = uni_app::App::new(uni_app::AppConfig {
        size: (800, 600),
        title: "my game".to_owned(),
        vsync: true,
        show_cursor: true,
        headless: false,
        resizable: true,
        fullscreen: false,
        intercept_close_request: false,
    });
    // retrieve the opengl/webgl/webgl2 context
    let gl = uni_gl::WebGLRenderingContext::new(app.canvas());

    // now do some opengl stuff
    // version of the shader depends if we are native or embedded
    let version = if uni_gl::IS_GL_ES { "300 es" } else { "150" };
    let vert_shader = compile_shader(
        &gl,
        uni_gl::ShaderKind::Vertex,
        &format!(
            r##"#version {version}
            in vec4 position;
            void main() {{
                gl_Position = position;
            }}
        "##
        ),
    );

    let frag_shader = compile_shader(
        &gl,
        uni_gl::ShaderKind::Fragment,
        &format!(
            r##"#version {version}
            precision mediump float;
            uniform float time;
            out vec4 FragColor;
            void main() {{
                float r=(cos(time)+1.0)*0.5;
                float g=(sin(time)+1.0)*0.5;
                FragColor = vec4(r,g,0.5,1);
            }}
        "##,
        ),
    );
    let program = link_program(&gl, &vert_shader, &frag_shader);
    gl.use_program(&program);

    let vertices: Vec<f32> = vec![-0.7, -0.7, 0.0, 0.7, -0.7, 0.0, 0.0, 0.7, 0.0];
    let vert_count = vertices.len() / 3;
    let position_attribute_location = gl.get_attrib_location(&program, "position").unwrap();
    let buffer = gl.create_buffer();
    gl.bind_buffer(uni_gl::BufferKind::Array, &buffer);
    gl.buffer_data(
        uni_gl::BufferKind::Array,
        &vertices.into_bytes(),
        uni_gl::DrawMode::Static,
    );
    let vao = gl.create_vertex_array();
    gl.bind_vertex_array(&vao);
    gl.vertex_attrib_pointer(
        position_attribute_location,
        uni_gl::AttributeSize::Three,
        uni_gl::DataType::Float,
        false,
        0,
        0,
    );
    gl.enable_vertex_attrib_array(position_attribute_location);
    gl.bind_vertex_array(&vao);

    // start game loop
    app.run(move |_app: &mut uni_app::App| {
        let now = uni_app::now();
        let time = gl.get_uniform_location(&program, "time").unwrap();
        gl.uniform_1f(&time, now as f32);
        gl.clear_color(0.0, 0.0, 0.0, 1.0);
        gl.clear(uni_gl::BufferBit::Color);
        // render a triangle
        gl.draw_arrays(uni_gl::Primitives::Triangles, vert_count);
    });
}

pub fn compile_shader(
    gl: &uni_gl::WebGLRenderingContext,
    shader_type: uni_gl::ShaderKind,
    source: &str,
) -> uni_gl::WebGLShader {
    let shader = gl.create_shader(shader_type);
    gl.shader_source(&shader, source);
    gl.compile_shader(&shader);
    shader
}

pub fn link_program(
    gl: &uni_gl::WebGLRenderingContext,
    vert_shader: &uni_gl::WebGLShader,
    frag_shader: &uni_gl::WebGLShader,
) -> uni_gl::WebGLProgram {
    let program = gl.create_program();
    gl.attach_shader(&program, vert_shader);
    gl.attach_shader(&program, frag_shader);
    gl.link_program(&program);
    program
}
